basic: Add log context

DaanDeMeyer opened the pull request:
This commit adds support for attaching extra metadata to log
messages written to the journal via log.h. We keep track of a
thread local log context in log.c onto which we can push extra
metadata fields that should be logged. Once a field is no longer
relevant, it can be popped again from the log context.

On top of this, we then add macros to allow pushing extra fields
onto the log context.

LOG_CONTEXT_PUSH() will push the provided field onto the log context
and pop the last field from the log context when the current block
ends. LOG_CONTEXT_PUSH_STRV() will do the same but for all fields in
the given strv.

Using the macros is as simple as putting them anywhere inside a block
to add a field to all following log messages logged from inside that
block.

```
void myfunction(...) {
	...

	LOG_CONTEXT_PUSH("MYMETADATA=abc");

	// Every journal message logged will now have the MYMETADATA=abc 
        // field included.
}
```

One special case to note is async code, where we use callbacks that
are invoked to continue processing when some event occurs. For async
code, there's usually an associated "userdata" struct containing
all the information associated with the async request. In this
"userdata" struct, we can add an extra "log_fields" field which is
a strv containing extra fields to be attached to log messages. At
the start of each callback, we can then call
"LOG_CONTEXT_PUSH_STRV(userdata->log_fields)" which makes sure the
fields are attached for the duration of the callback.

bluca commented on src/basic/log.c:
```diff
@@ -617,15 +623,15 @@ static int write_to_journal(
+        log_context[1] = IOVEC_MAKE_STRING("MESSAGE=");
+        log_context[2] = IOVEC_MAKE_STRING(buffer);
+        log_context[3] = IOVEC_MAKE_STRING("\n");
+        /* Make sure to check if LOG_CONTEXT_RESERVED is large enough when adding more fields here. */
```
can this be done with a cc_assert?

DaanDeMeyer commented:
Removed the CONSUME versions of the macros as they didn't allow for proper OOM handling.

DaanDeMeyer commented:
Reworked implementation to make sure we add the extra metadata everywhere. All stack allocated iovecs are now replaced by use of the log context iovec array to avoid having to copy iovecs around.

poettering commented:
Hmpf. I really dislike using TLS for things like this. TLS is costly, in particular if you stick an array in it, and releasing it is cumbersome.

I mean, i see why this is useful, but can't we do this without tls? for example using block based stack allocations. i.e. have something like this:


```c
struct log_context {
        const char *field;
        struct log_context *prev;
};

#define LOG_CONTEXT_BEGIN \
        struct log_context *_log_context = NULL;

#define LOG_CONTEXT_ADD(s) \
        _log_context = &(struct log_context) { .field = s, .prev = _log_context }
```

Then you could do: 

```c
int foo(void) {
        LOG_CONTEXT_BEGIN;

        LOG_CONTEXT_ADD("foo=bar");
        LOG_CONTEXT_ADD("wuff=kkk");

        log_info("ydfkjhdlkfhdskjf"); /* this macro would then have to pick up _log_context as parameter implicitly */
}
```

next step would then be to change LOG_CONTEXT_ADD() to also use `_cleanup_` stuff to automatically remove an item from the list when the current block is left, so that this would work:

```c
int foo(void) {
        LOG_CONTEXT_BEGIN;

        LOG_CONTEXT_ADD("foo=bar");
        LOG_CONTEXT_ADD("wuff=kkk");

        for (;;) {
                _cleanup_free_ char *p = NULL;
                asprintf(&p, "random=%" PRIx64, random_u64());
                LOG_CONTEXT_ADD(p);
                
                log_info("something happened");
        }
```

poettering commented:
Hmpf. I really dislike using TLS for things like this. TLS is costly, in particular if you stick an array in it, and releasing it is cumbersome.

I mean, i see why this is useful, but can't we do this without tls? for example using block based stack allocations. i.e. have something like this:


```c
struct log_context {
        const char *field;
        struct log_context *prev;
};

#define LOG_CONTEXT_BEGIN \
        struct log_context *_log_context = NULL;

#define LOG_CONTEXT_ADD(s) \
        _log_context = &(struct log_context) { .field = s, .prev = _log_context }
```

Then you could do: 

```c
int foo(void) {
        LOG_CONTEXT_BEGIN;

        LOG_CONTEXT_ADD("foo=bar");
        LOG_CONTEXT_ADD("wuff=kkk");

        log_info("ydfkjhdlkfhdskjf"); /* this macro would then have to pick up _log_context as parameter implicitly */
}
```

next step would then be to change LOG_CONTEXT_ADD() to also use `_cleanup_` stuff to automatically remove an item from the list when the current block is left, so that this would work:

```c
int foo(void) {
        LOG_CONTEXT_BEGIN;

        LOG_CONTEXT_ADD("foo=bar");
        LOG_CONTEXT_ADD("wuff=kkk");

        for (;;) {
                _cleanup_free_ char *p = NULL;
                asprintf(&p, "random=%" PRIx64, random_u64());
                LOG_CONTEXT_ADD(p);
                
                log_info("something happened");
        }
```

poettering commented:
i'd be more OK with using TLS if we only store the anchor of such a linked list in it... i.e. singular fields that require no cleanup should be ok.

poettering commented:
i'd be more OK with using TLS if we only store the anchor of such a linked list in it... i.e. singular fields that require no cleanup should be ok.

DaanDeMeyer commented:
Hmm, that's a neat approach, let me try that out, I think it should work.

DaanDeMeyer commented:
Re-implemented using the approach suggested by @poettering. PTAL!

bluca commented on src/basic/log.h:
```diff
@@ -367,3 +367,37 @@ int log_syntax_invalid_utf8_internal(
+extern thread_local struct log_context *_log_context;
+
+static inline void _log_context_pop(struct log_context **c) {
+        if (!*c)
```
`if (!c || !*c)`

DaanDeMeyer replied:
We don't do this check either in `DEFINE_TRIVIAL_CLEANUP_FUNC()` so I left it out here as well. I think it's guaranteed c itself won't be `NULL`?

bluca replied:
force of habit... doesn't cost much to check, and it's a bit friendlier to onlookers who might not know what is guaranteed and what is not - in fact if it's guaranteed, then an assert() to assert it would be a nice way to document it clearly

bluca commented on src/test/test-log.c:
```diff
@@ -70,6 +70,25 @@ static void test_log_syntax(void) {
+                LOG_CONTEXT_PUSH("MYDATA=abc");
+                LOG_CONTEXT_PUSH("MYDATA=def");
+                LOG_CONTEXT_PUSH_STRV(strv);
+                LOG_CONTEXT_PUSH_STRV(strv);
```
Should it work if calling a function after this line, and doing more LOG_CONTEXT_PUSH from there? If so, would probably be good to cover that case too in the test

DaanDeMeyer replied:
Extended the test a bit more in general. Went for a nested block since that's pretty much the same as a function in this scenario.

bluca commented on src/basic/log.h:
```diff
@@ -367,3 +368,37 @@ int log_syntax_invalid_utf8_internal(
 #define DEBUG_LOGGING _unlikely_(log_get_max_level() >= LOG_DEBUG)
 
 void log_setup(void);
+
```
Maybe add a brief comment here, explaining what's this useful for, and how to use it (maybe explicitly mention that the caller owns any memory buffer that gets pushed, and popping doesn't free them)?

DaanDeMeyer replied:
Added a blurb

keszybz commented on src/basic/log.c:
```diff
@@ -607,15 +632,19 @@ static int write_to_journal(
         log_do_header(header, sizeof(header), level, error, file, line, func, object_field, object, extra_field, extra);
 
-        struct iovec iovec[4] = {
+        struct iovec iovec[JOURNAL_MAX_ENTRIES] = {
```
Does JOURNAL_MAX_ENTRIES need to be so big? `struct iovec` is 16 bytes, so we're allocating 16k here on the stack.

DaanDeMeyer replied:
I copied the max amount of entries that a single journal message can contain. We can certainly lower this though. I'll put in a lower value

DaanDeMeyer replied:
Lowered to 128. This gives us space for 64 fields

keszybz commented on src/basic/log.c:
```diff
@@ -588,6 +592,27 @@ static int log_do_header(
+        assert(iovec);
+        assert(n);
+
+        if (iovec_len <= 1)
+                return;
```
Is this check needed? We don't expect such short iovec, and the check below should handle it if we get one.

keszybz commented on src/basic/log.c:
```diff
@@ -588,6 +592,27 @@ static int log_do_header(
+
+        LIST_FOREACH(ll, c, _log_context) {
+                STRV_FOREACH(s, _log_context->fields) {
+                        if (*n >= iovec_len - 1)
+                                break;
+
+                        iovec[(*n)++] = IOVEC_MAKE_STRING(*s);
+                        iovec[(*n)++] = IOVEC_MAKE_STRING("\n");
+                }
+
+                if (*n >= iovec_len - 1)
+                        break;
```
This repeated check is annoying. Why not just drop the second check and `return` from the first? We don't expect any items in this list with empty fields, so we don't need to optimize for that.

keszybz commented on src/basic/log.c:
```diff
@@ -588,6 +592,27 @@ static int log_do_header(
+                return;
+
+        LIST_FOREACH(ll, c, _log_context) {
+                STRV_FOREACH(s, _log_context->fields) {
```
I'm afraid this needs to be `c->fields` ;(

Maybe add a test that this logging actually works in TEST-04-JOURNAL ?

DaanDeMeyer replied:
Hmm, let me extend systemd-cat with support for adding extra fields and use this new functionality there immediately. Then I can use that in TEST-04-JOURNAL to test that this works.

keszybz commented on src/basic/log.h:
```diff
@@ -367,3 +368,68 @@ int log_syntax_invalid_utf8_internal(
+        LIST_REMOVE(ll, _log_context, *c);
+}
+
+#define _LOG_CONTEXT_PUSH(s, c)                                                     \
+        _cleanup_(_log_context_pop) struct log_context *c = &(struct log_context) { \
+                .fields = (char*[]) { (char *) (s), NULL },                         \
+        };                                                                          \
+        LIST_PREPEND(ll, _log_context, c)
+
+#define LOG_CONTEXT_PUSH(s) \
+        _LOG_CONTEXT_PUSH(s, UNIQ_T(c, UNIQ))
+
+#define _LOG_CONTEXT_PUSH_STRV(strv, c)                                             \
+        _cleanup_(_log_context_pop) struct log_context *c = &(struct log_context) { \
+                .fields = (strv),                                                   \
+        };                                                                          \
+        LIST_PREPEND(ll, _log_context, c)
```
Hmm, do we need the separate version for strv? Do you expect any uses where the strv would be passed from outside?
I'd imagine that the context is always constructed locally, so we could have just a single macro LOG_CONTEXT_PUSH() which accepts one or more strings and calls STRV_MAKE() on them? (Also, STRV_MAKE should be used above…)

DaanDeMeyer replied:
I explained this a bit in the comment, but I figure if we want to use this for e.g. a resolved DNS request, and we want to retain context across asynchronous calls, we'd gather the context fields in a strv, store that strv in some `userdata` struct and push it onto the context at the start of each callback associated with that DNS request. That's the primary use case I had in mind for this macro.

keszybz replied:
Ah, OK. I remember reading that, but it was a while ago… But in that case, shouldn't LOG_CONTEXT_PUSH just be a wrapper for `LOG_CONTEXT_PUSH_STRV(STRV_MAKE(__VARARGS__))` ?

DaanDeMeyer replied:
That makes sense, I'll rework this to do that

poettering commented on src/test/test-log.c:
```diff
@@ -70,6 +70,54 @@ static void test_log_syntax(void) {
 }
 
+static void test_log_context(void) {
+        char *strv[] = { (char*) "MYDATA=abc", NULL };
```
STRV_MAKE?

maybe add multiple entries here, to actually use the fact an strv is accepted here

poettering commented on src/basic/log.c:
```diff
@@ -32,12 +32,14 @@
 #include "utf8.h"
 
 #define SNDBUF_SIZE (8*1024*1024)
+#define MAX_IOVEC 128
```
usually we put the MAX last, i.e. IOVEC_MAX?

poettering replied:
btw, why do this? why not instead a counter also as thread_local variable that encodes how many entries are needed? it's increased whenever a context is added, and decreased whenever one is removed.

DaanDeMeyer replied:
To avoid unbounded stack allocations, I figure it makes sense to impose an upper limit on the amount of stuff we put into the iovec. If we already have a max size, tracking the size of the log context linked list individually doesn't seem like it provides much benefit

poettering commented:
looks ok to me, just some minor points...

yuwata commented on src/basic/list.h:
```diff
@@ -1,8 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once
 
-#include "macro.h"
```
Why this is dropped?

DaanDeMeyer replied:
We get compilation failures if we leave it in:

```
➜  systemd git:(log-context) ninja -C build
ninja: Entering directory `build'
[165/2017] Compiling C object src/libsystemd/libsystemd_static.a.p/sd-netlink_netlink-types.c.o
FAILED: src/libsystemd/libsystemd_static.a.p/sd-netlink_netlink-types.c.o 
cc -Isrc/libsystemd/libsystemd_static.a.p -Isrc/libsystemd -I../src/libsystemd -Isrc/basic -I../src/basic -Isrc/fundamental -I../src/fundamental -Isrc/systemd -I../src/systemd -I. -I.. -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-journal -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=gnu11 -O0 -g -Wno-format-signedness -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wno-unused-result -Werror=missing-declarations -Werror=missing-prototypes -fdiagnostics-show-option -fno-common -fno-strict-aliasing -fstack-protector -fstack-protector-strong -fvisibility=hidden --param=ssp-buffer-size=4 -fno-omit-frame-pointer -Werror=shadow -include config.h -fPIC -pthread -fvisibility=default -MD -MQ src/libsystemd/libsystemd_static.a.p/sd-netlink_netlink-types.c.o -MF src/libsystemd/libsystemd_static.a.p/sd-netlink_netlink-types.c.o.d -o src/libsystemd/libsystemd_static.a.p/sd-netlink_netlink-types.c.o -c ../src/libsystemd/sd-netlink/netlink-types.c
In file included from ../src/basic/macro.h:452,
                 from ../src/basic/list.h:4,
                 from ../src/libsystemd/sd-netlink/netlink-internal.h:8,
                 from ../src/libsystemd/sd-netlink/netlink-types.c:6:
../src/basic/log.h:405:9: error: expected specifier-qualifier-list before ‘LIST_FIELDS’
  405 |         LIST_FIELDS(struct log_context, ll);
      |         ^~~~~~~~~~~
../src/basic/log.h:408:51: error: unknown type name ‘_log_context’
  408 | extern thread_local LIST_HEAD(struct log_context, _log_context);
      |                                                   ^~~~~~~~~~~~
../src/basic/log.h: In function ‘_log_context_pop’:
../src/basic/log.h:416:9: error: implicit declaration of function ‘LIST_REMOVE’ [-Werror=implicit-function-declaration]
  416 |         LIST_REMOVE(ll, _log_context, *c);
      |         ^~~~~~~~~~~
../src/basic/log.h:416:9: warning: nested extern declaration of ‘LIST_REMOVE’ [-Wnested-externs]
../src/basic/log.h:416:21: error: ‘ll’ undeclared (first use in this function)
  416 |         LIST_REMOVE(ll, _log_context, *c);
      |                     ^~
../src/basic/log.h:416:21: note: each undeclared identifier is reported only once for each function it appears in
../src/basic/log.h:416:25: error: ‘_log_context’ undeclared (first use in this function); did you mean ‘log_context’?
  416 |         LIST_REMOVE(ll, _log_context, *c);
      |                         ^~~~~~~~~~~~
      |                         log_context
cc1: some warnings being treated as errors
[174/2017] Compiling C object src/libsystemd/libsystemd_static.a.p/sd-resolve_sd-resolve.c.o
ninja: build stopped: subcommand failed.
```

yuwata replied:
How about to include macro.h at the end of the file, like macro.h includes logs.h? Otherwise, we cannot safely use the macros defined in list.h without separately including macro.h.

yuwata commented on src/basic/log.c:
```diff
@@ -588,6 +592,21 @@ static int log_do_header(
+        assert(iovec);
+        assert(n);
+
+        LIST_FOREACH(ll, c, _log_context) {
```
You can drop `{`.

yuwata commented on src/basic/log.c:
```diff
@@ -588,6 +592,21 @@ static int log_do_header(
+
+        LIST_FOREACH(ll, c, _log_context) {
+                STRV_FOREACH(s, c->fields) {
+                        if (*n >= iovec_len - 1)
```
This should be `-2`??

yuwata commented on src/basic/log.c:
```diff
@@ -1045,16 +1071,24 @@ int log_struct_iovec_internal(
-                struct iovec iovec[1 + n_input_iovec*2];
-                iovec[0] = IOVEC_MAKE_STRING(header);
+                /* Each input iovec translates to 2 iovecs in our iovec array because we need to add a newline. */
+                if (n_input_iovec > IOVEC_MAX / 2)
```
`if (n_input_iovec * 2 + 1 > IOVEC_MAX)`??

DaanDeMeyer replied:
I went for `n_input_iovec * 2 >= IOVEC_MAX)` which should be equivalent

yuwata commented on src/basic/log.h:
```diff
@@ -8,6 +8,7 @@
 #include <syslog.h>
 
 #include "macro.h"
+#include "list.h"
```
Please sort the headers.

yuwata commented on src/basic/log.h:
```diff
@@ -367,3 +368,62 @@ int log_syntax_invalid_utf8_internal(
+}
+
+#define LOG_CONTEXT_PUSH(...) \
+        LOG_CONTEXT_PUSH_STRV(STRV_MAKE(__VA_ARGS__))
```
CONST?

yuwata reviewed (approved):
LGTM.

poettering commented:
So this PR currently doesn't port anything over to the new interface. Can you maybe add that? We should generally have at least one user of the stuff we have here.

DaanDeMeyer commented:
Reworked the interface a bit, and added a few things that I figured out would be nice while trying to convert some code to use this. Still need to push the code that makes some part of the codebase use this feature.

- Added LOG_CONTEXT_PUSHF() to add a field via formatting
- We now expose the underlying log context struct (renamed to LogContext) to make the entire feature more flexible overall. For example, we can now write a function `log_context_push_device()` that creates a new log context containing all the fields relevant to a `sd_device` instance.

yuwata commented on src/basic/stdio-util.h:
```diff
@@ -9,13 +9,18 @@
-                int _snpf = snprintf(_buf, _len, (fmt), __VA_ARGS__);  \
-                _snpf >= 0 && (size_t) _snpf < _len ? _buf : NULL;     \
-        })
+static inline char *snprintf_ok(char *buf, size_t len, const char *format, ...) {
```
I think it is better to add `_printf_` attribute.

yuwata commented on src/basic/log.c:
```diff
@@ -1478,3 +1511,35 @@ void log_setup(void) {
+        if (!c)
+                return NULL;
+
+        *c = (LogContext) {};
```
Please use `new0()`.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +114,83 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+        if (r < 0 && r != -ENOENT)
+                log_error_errno(r, "Failed to get device \"%s\" property: %m", field);
+
+        if (r == 0)
```
Please use `r >= 0`.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +114,83 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+                sd_device *device,
+                const char *field,
+                int (*func)(sd_device *dev, const char **s),
+                char ***ret) {
```
`ret` here is also an input. Please rename it.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +114,83 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+
+        assert(device);
+        assert(field);
+        assert(func);
```
Missing assertion for `ret`.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +114,83 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+
+        r = func(device, &s);
+        if (r < 0 && r != -ENOENT)
+                log_error_errno(r, "Failed to get device \"%s\" property: %m", field);
```
Please downgrade error and mention that the failure is ignored.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +114,83 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+                log_error_errno(r, "Failed to get device \"DEVNUM\" property: %m");
+
+        if (r == 0)
+                (void) strv_extendf(&strv, "DEVNUM=%u:%u", major(devnum), minor(devnum));
```
Nowadays we have macro for formatting devnum.

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+
+        r = func(device, &s);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"%s\" property, ignoring: %m", field);
```
`log_device_debug_errno()`?

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"DEVNUM\" property, ignoring: %m");
```
Also here, `log_device_debug_errno()` is better?

yuwata commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+        if (r >= 0)
+                (void) strv_extendf(&strv, "DISKSEQ=%"PRIu64, diskseq);
+
+        return TAKE_PTR(strv);
```
How about to cash the strv in sd-device object? Then, we can enable/disable log context frequently with almost no performance regression.

DaanDeMeyer replied:
See the other discussion, I don't think we'll be enabling/disabling the log context very frequently, so I'd like to leave this optimization until we find a use case where we're enabling/disabling the same log context very frequently.

yuwata commented on src/libsystemd/sd-device/device-monitor.c:
```diff
@@ -245,6 +245,8 @@ static int device_monitor_event_handler(sd_event_source *s, int fd, uint32_t rev
         if (device_monitor_receive_device(m, &device) <= 0)
                 return 0;
 
+        LOG_CONTEXT_CONSUME_STRV(device_log_context_fields(device));
```
Setting the log context here may be too early, as an uevent may trigger very long operation.

DaanDeMeyer replied:
I don't understand, this is right before we invoke the callback no? If this is too early, where would you suggest we set the log context for the device?

yuwata replied:
In other words, the usage of log context I thought was that each sd-device object manages its log context for the device, and push and pop within single call of `log_device_xxx()`. Then, log messages unrelated to the device are not polluted by the log context.

But now, you propose all logs during a uevent attaches the log context about the received device. That might help in some case. But when we widely use the log context for variety of objects, then each log messages has toooo many fields...

DaanDeMeyer replied:
The device log context is only valid for the duration of the callback though. And I figure that during the callback for a specific device, it makes sense to attach the log context fields related to the device to every log message logged during the callback.

This makes it very easy to show all log messages related to the device by simply doing `journalctl DEVNAME=<devname>`.

The idea is that during a device monitor callback for a device X, even when we're calling into code that isn't aware of the device being processed, we can still attach the necessary context to those log messages to be able to figure out later that they were logged when device X was being processed.

To be able to only show the fields you're interested in, I've been working on https://github.com/systemd/systemd/pull/23635 which will allow to format the journal output to only show the fields the user is interested in.

As for disk space constraints, the journal deduplicates identical fields so logging the same fields many times for many log messages is cheap.

DaanDeMeyer commented on src/libsystemd/sd-bus/sd-bus.c:
```diff
@@ -2933,6 +2933,7 @@ static int process_message(sd_bus *bus, sd_bus_message *m) {
         bus->current_message = m;
         bus->iteration_counter++;
 
+        LOG_CONTEXT_CONSUME_STRV(bus_message_log_context_fields(m));
```
This doesn't only apply to the debug log message below. During `process_message()`, all the callbacks registered in the vtable for the service get called as well.

The idea of adding the fields to the context here is to add these fields to any log message that's logged from those callbacks, so that later, we can easily query for all the log messages that happened during processing of a specific D-Bus method by adding a `DBUS_METHOD=Reload` match to journalctl for example.

poettering commented on src/basic/log.c:
```diff
@@ -588,6 +592,20 @@ static int log_do_header(
+
+        LIST_FOREACH(ll, c, _log_context)
+                STRV_FOREACH(s, c->fields) {
+                        if (*n >= iovec_len - 2)
```
might underflow if iovec_len is smaller than 2?

add assert? or other way to handle corner case?

poettering commented on src/basic/log.c:
```diff
@@ -607,15 +625,19 @@ static int write_to_journal(
         log_do_header(header, sizeof(header), level, error, file, line, func, object_field, object, extra_field, extra);
 
-        struct iovec iovec[4] = {
+        struct iovec iovec[IOVEC_MAX] = {
```
that's quite an increase. i.e. 32 byte before. 1K now. in a world where 4K stacks are often used... For a log function that should try to be as little intrusive as we can be.

maybe count how much we actually need, and size the thing properly?

DaanDeMeyer replied:
Added counting, but kept the upper limit just in case

poettering commented on src/basic/log.h:
```diff
@@ -367,3 +369,68 @@ int log_syntax_invalid_utf8_internal(
+ */
+
+typedef struct LogContext {
+        char **fields;
```
Can you add a comment here: `/* depending on which destructor is used (log_context_free() or log_context_detach()) the memory referenced by this is freed or not */` or so?

If there are two ways memory is owned there should always be a specific note about it where the field for it is declared.

poettering commented on src/basic/stdio-util.h:
```diff
@@ -8,6 +8,7 @@
 #include "macro.h"
 
+_printf_(3, 4)
 static inline char *snprintf_ok(char *buf, size_t len, const char *format, ...) {
```
should probably be a separate commit

poettering commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0 && r != -ENOENT)
+                log_device_debug_errno(device, r, "Failed to get device \"DEVNUM\" property, ignoring: %m");
+
```
drop this empty line given you are still processing the same fuction return code here

poettering commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+        r = sd_device_get_ifindex(device, &ifindex);
+        if (r < 0 && r != -ENOENT)
+                log_device_debug_errno(device, r, "Failed to get device \"IFINDEX\" property, ignoring: %m");
+
```
ditto and everywhere else

poettering commented on src/libsystemd/sd-device/device-util.c:
```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+        return 0;
+}
+
+char** device_log_context_fields(sd_device *device) {
```
make call this `device_make_log_fields()`?

poettering commented on src/basic/log.h:
```diff
@@ -367,3 +369,68 @@ int log_syntax_invalid_utf8_internal(
+        _unused_ _cleanup_(log_context_freep) LogContext *c = log_context_new(strv);
+
+#define LOG_CONTEXT_CONSUME_STRV(strv) \
+        _LOG_CONTEXT_CONSUME_STRV(strv, UNIQ_T(c, UNIQ))
```
we really should have `LOG_CONTET_CONSUME_STR()` too, no? how would we do that?

DaanDeMeyer replied:
Added `LOG_CONTEXT_CONSUME_STR()` using `strv_new()`.

Maybe we should have `strv_ops` that we can pass to `strv_free()` to indicate how the strv should be freed.

poettering commented on src/test/test-log.c:
```diff
@@ -70,6 +71,85 @@ static void test_log_syntax(void) {
+        }
+
+        {
+                char **strv = strv_new("ABC", "DEF");
```
assert_se()

poettering commented on src/test/test-log.c:
```diff
@@ -70,6 +71,85 @@ static void test_log_syntax(void) {
+        assert(_log_context == NULL);
+
+        {
+                _cleanup_(log_context_freep) LogContext *ctx = log_context_new(NULL);
```
assert_se()

poettering commented on src/libsystemd/sd-bus/bus-message.c:
```diff
@@ -4649,3 +4649,42 @@ _public_ int sd_bus_message_sensitive(sd_bus_message *m) {
+        assert(m);
+
+        if ((s = bus_message_type_to_string(m->header->type)))
+                (void) strv_extendf(&strv, "DBUS_MESSAGE_TYPE=%s", s);
```
maybe let's add to strv.h a helper:

```c
int strv_extend_assignment(char ***l, const char *lhs, const char *rhs) {
        assert(l);
        assert(lhs);

        if (!rhs) /* value is optional, in which case we suppress the field */
                  return 0;
    
        j = strjoin(lhs, "=", "rhs");
        if (!j)
                return NULL;

         return strv_consume(l, j); 
}
``

poettering commented on src/basic/strv.c:
```diff
@@ -904,6 +904,22 @@ int strv_extend_n(char ***l, const char *value, size_t n) {
         return -ENOMEM;
 }
 
+int strv_extend_assignment(char ***l, const char *lhs, const char *rhs) {
```
this smells a lot like strv_env_assign()...

I wonder if it wouldn't be better to make this new helper more generic, i.e. call this strv_extend_join() that combined strjoin() + strv_extend(). You'd thus call:

```c
r = strv_extend_join(&l, "foo", "=", "bar");
```

I think that would be almost as good to read, less redundant given that strv_env_assign() already exists, and I am sure we can find various other places to use this.

poettering commented on src/basic/list.h:
```diff
@@ -190,3 +188,5 @@
                 _p;                                                     \
         })
+
+#include "macro.h"
```
hmm, this feels like a hack, but i got no better idea atm.

But please add a comment why this is needed. i.e something like this:

```c
/* Now include "macro.h", because we want our definition of assert() which the macros above use. We include it down here instead of up top, since macro.h pulls in log.h which in turn needs our own definitions. */
```

or something like that

yuwata replied:
(BTW, we already do that in macro.h, i.e. log.h is included at the end of macro.h)

poettering commented on src/basic/log.c:
```diff
@@ -588,6 +592,21 @@ static int log_do_header(
+
+        LIST_FOREACH(ll, c, _log_context)
+                STRV_FOREACH(s, c->fields) {
+                        if (*n >= iovec_len - 2)
```
maybe write as `if (*n + 2 >= iovec_len)` to avoid the underflow and the assert check?

poettering commented on src/basic/log.c:
```diff
@@ -1478,3 +1516,47 @@ void log_setup(void) {
+                        num_fields++;
+
+        return num_fields;
+}
```
hmm, if we have large numbers of fields this likely will get quite expensive... a static counter next to the linked list might be more future proof there? This function is called on every log message after all

poettering commented on src/basic/log.c:
```diff
@@ -1478,3 +1516,47 @@ void log_setup(void) {
+}
+
+LogContext* log_context_new(char **fields) {
+        LogContext *c = new(LogContext, 1);
```
maybe add an `strv_isempty()` shortcut to bypass things?

poettering commented on src/basic/log.h:
```diff
@@ -406,3 +408,79 @@ typedef struct LogRateLimit {
+ */
+
+#define _LOG_CONTEXT_CONSUME_STR(s, c) \
+        _unused_ _cleanup_(log_context_freep) LogContext *c = log_context_new(strv_new(s))
```
this will leak memory if the strv_new() succeeds but the log_context_new() does not

poettering commented on src/libsystemd/sd-device/device-monitor.c:
```diff
@@ -246,6 +246,8 @@ static int device_monitor_event_handler(sd_event_source *s, int fd, uint32_t rev
         if (device_monitor_receive_device(m, &device) <= 0)
                 return 0;
 
+        LOG_CONTEXT_CONSUME_STRV(device_make_log_fields(device));
```
hmm, i am not sure i like this use here...

I mean, think about the callback function called below might start doing pretty much anything, for example a screen redraw in a GUI app. But if some graphics function then logs, should that really carry device info? I am not sure...

Using this funcitonality is good, but when doing a generic callback dispatch then that's going to far I think, because you have no clue what the callback will do and if it will issue more than just stuff directly related to the device even we have here...

DaanDeMeyer replied:
Maybe this is excessive information to attach by default. But it's still immensely useful information to know that the graphics function logged a message during the processing of a device monitor callback for device X.

You can go pretty wild with the context concept. If you want to be able to filter all log messages filtered during the execution of a particular function, just push some unique field on the context and you can filter by that in journalctl.

yuwata replied:
Still I do not like this. And I agree with @poettering's concern in the above. Relevant discussion is here https://github.com/systemd/systemd/pull/23309#discussion_r947240763 (sorry for late response, though).

poettering commented on src/libsystemd/sd-device/device-util.h:
```diff
@@ -83,3 +83,6 @@
 int device_open_from_devnum(mode_t mode, dev_t devnum, int flags, char **ret);
+
+char** device_make_log_fields(sd_device *device);
+
```
trailing empty line

poettering commented on src/libsystemd/sd-bus/bus-message.c:
```diff
@@ -4649,3 +4649,26 @@ _public_ int sd_bus_message_sensitive(sd_bus_message *m) {
+        (void) strv_extend_assignment(&strv, "DBUS_MEMBER", sd_bus_message_get_member(m));
+
+        (void) strv_extendf(&strv, "DBUS_MESSAGE_COOKIE=" PRIu64, BUS_MESSAGE_COOKIE(m));
+        if (m->reply_cookie)
```
coding style: this is not boolean/boolean-like or a pointer, but a numeric integer, hence check with `reply->cookie != 0`

DaanDeMeyer commented on src/basic/stdio-util.h:
```diff
@@ -9,13 +9,19 @@
+        va_list ap;
+        int r;
+
+        va_start(ap, format);
```
I want to pass a pre-allocated buffer via a compound initializer. If we have the macro + statement expressions, the compound initializer block will be scoped to the macro block. If we make it a static inline function, it will be scoped to the block of the function calling snprintf_ok().

keszybz replied:
Can you add this rationale to the commit description?

>  hmm, i wonder how the generated code here actually compares to the status quo ante?

That's a good question, but since we only have 30 invocations of snprintf_ok in the codebase, we don't really care if it's particularly efficient.

yuwata commented:
Or, maybe, instead of adding full device information, how about to add basic callback information e.g. `SYSTEMD_CALLBACK="sd-device-monitor(description)=sysname"`?

DaanDeMeyer commented:
I dropped the device stuff, we can discuss it in a separate PR, now only the D-Bus context fields are left

keszybz commented on src/basic/stdio-util.h:
```diff
@@ -7,7 +7,6 @@
 #include <sys/types.h>
 
 #include "macro.h"
-#include "memory-util.h"
```
Please add rationale to the commit message.

keszybz commented on src/libsystemd/sd-bus/bus-message.c:
```diff
@@ -4667,3 +4667,26 @@ _public_ int sd_bus_message_sensitive(sd_bus_message *m) {
+
+        assert(m);
+
+        (void) strv_extend_join(&strv, "DBUS_MESSAGE_TYPE", "=", bus_message_type_to_string(m->header->type));
+        (void) strv_extend_join(&strv, "DBUS_SENDER", "=", sd_bus_message_get_sender(m));
+        (void) strv_extend_join(&strv, "DBUS_DESTINATION", "=", sd_bus_message_get_destination(m));
+        (void) strv_extend_join(&strv, "DBUS_PATH", "=", sd_bus_message_get_path(m));
+        (void) strv_extend_join(&strv, "DBUS_INTERFACE", "=", sd_bus_message_get_interface(m));
+        (void) strv_extend_join(&strv, "DBUS_MEMBER", "=", sd_bus_message_get_member(m));
```
Hmm, would `strv_extend_join` ever be used with something else than "=" as the joiner? It'd be more effiecient to change those calls to
```C
 (void) strv_extend_join(&strv, "DBUS_MESSAGE_TYPE=", bus_message_type_to_string(m->header->type));
...
```

keszybz reviewed (commented):
I wasn't following the previous discussion, so apologies if I'm asking something that is already answered in some buried comment… Partial review, I didn't look at commit 5/6 that does the actual implementation yet.

keszybz reviewed (commented):
Looks good otherwise.

yuwata commented:
Sorry, but I do not like the last commit, as the same reason https://github.com/systemd/systemd/pull/23309#discussion_r965907330.

keszybz commented:
> Sorry, but I do not like the last commit, as the same reason [#23309 (comment)](https://github.com/systemd/systemd/pull/23309#discussion_r965907330).

OK, so you are worried that other applications using libsystemd can do arbitrary things in nested calls from libsystemd and that this metadata would be unwanted there. I guess this makes sense. @DaanDeMeyer what about the adding a flag for opting in? I.e. have some (unexported at least for now) function bus_enable_log_context() and systemd programs would call that and get the context logs after allocating the bus object.

DaanDeMeyer commented:
> OK, so you are worried that other applications using libsystemd can do arbitrary things in nested calls from libsystemd and that this metadata would be unwanted there. I guess this makes sense. @DaanDeMeyer what about the adding a flag for opting in? I.e. have some (unexported at least for now) function bus_enable_log_context() and systemd programs would call that and get the context logs after allocating the bus object.

That's OK for me. @yuwata would this address your concerns?

yuwata commented:
> > OK, so you are worried that other applications using libsystemd can do arbitrary things in nested calls from libsystemd and that this metadata would be unwanted there. I guess this makes sense. @DaanDeMeyer what about the adding a flag for opting in? I.e. have some (unexported at least for now) function bus_enable_log_context() and systemd programs would call that and get the context logs after allocating the bus object.
> 
> That's OK for me. @yuwata would this address your concerns?

Yeah. @poettering WDYT?

DaanDeMeyer commented:
I added back the device monitor log context stuff as well with the same fix so that enabling the log context for the device monitor requires an extra (private) function call

poettering commented on src/basic/strv.c:
```diff
@@ -789,6 +789,22 @@ int strv_extend_n(char ***l, const char *value, size_t n) {
         return -ENOMEM;
 }
 
+int strv_extend_join(char ***l, const char *lhs, const char *rhs) {
```
hmm, one of these days this should be rework to take variadic params, and dispatch to some flavour of strjoin_real()

poettering commented on src/basic/strv.c:
```diff
@@ -789,6 +789,22 @@ int strv_extend_n(char ***l, const char *value, size_t n) {
+        assert(lhs);
+
+        if (!rhs) /* value is optional, in which case we suppress the field */
+                return 0;
```
hmm, these semantics sound quite specific to me. should this really be in strv?

at the very least it should have a different name, i.e. strv_extend_with_assignment() or so?

shouldn't the `=` be inserted automatically here?

poettering commented on src/basic/log.c:
```diff
@@ -68,6 +70,17 @@ static bool prohibit_ipc = false;
+} LogContext;
+
+thread_local LIST_HEAD(LogContext, _log_context) = NULL;
+thread_local size_t _log_context_num_fields = 0;
```
no `static`?

poettering commented on src/basic/log.c:
```diff
@@ -1477,3 +1523,72 @@ void log_setup(void) {
+        assert(c);
+
+        STRV_FOREACH(s, c->fields)
+                _log_context_num_fields++;
```
```c
_log_context_num_fields += strv_length(c->fields)
```

poettering commented on src/basic/log.c:
```diff
@@ -1477,3 +1523,72 @@ void log_setup(void) {
+                return NULL;
+
+        STRV_FOREACH(s, c->fields)
+                _log_context_num_fields--;
```
similar.

also, please add assert() checking for underflow ahead of time

poettering commented on src/basic/log.c:
```diff
@@ -1477,3 +1523,72 @@ void log_setup(void) {
+
+LogContext* log_context_new(char **fields, bool owned) {
+        if (strv_isempty(fields))
+                return NULL;
```
hmm, this makes it hard to distinguish allocation failures from invocations with an empty list.

maybe add:

```c
#define LOG_CONTEXT_EMPTY ((LogContext*) (uintptr_t) 1)
```

as a special value you can return here to indicate an empty log context.

And then handle that specially in log_context_attach(), log_context_free() and so on.

poettering replied:
or, alternatively, just define a static empty one as:

```c
static const LogContext empty_log_context = {};
```

and then return a pointer to that?

DaanDeMeyer replied:
I just removed the empty `fields` check, if there are no fields, we still allocate a log context object. We can't use a static const instance here because we always have to return an allocated object that the user can free.

poettering commented on src/basic/log.c:
```diff
@@ -1477,3 +1523,72 @@ void log_setup(void) {
+        return mfree(c);
+}
+
+LogContext* log_context_consume(char **fields) {
```
maybe name log_context_consume_new()?

poettering commented on src/libsystemd/sd-bus/bus-message.c:
```diff
@@ -4667,3 +4667,26 @@ _public_ int sd_bus_message_sensitive(sd_bus_message *m) {
+
+        assert(m);
+
+        (void) strv_extend_join(&strv, "DBUS_MESSAGE_TYPE=", bus_message_type_to_string(m->header->type));
```
I think strv_extend_join() shiuld really be called strv_append_assignment(), and then insert the `=` on its own

poettering commented on src/libsystemd/sd-bus/bus-message.c:
```diff
@@ -4667,3 +4667,26 @@ _public_ int sd_bus_message_sensitive(sd_bus_message *m) {
+        (void) strv_extend_join(&strv, "DBUS_ERROR_MESSAGE=", m->error.message);
+
+        return TAKE_PTR(strv);
+}
```
instead of enabling this manually in all our code, shouldn't this be something we just enable when `DEBUG_LOGGING` holds?

That way, once you enable debug logging you get the extra fields for *all* messages just like that, but without it we won't waste time on all those memory allocations?

poettering replied:
or here's another idea, instead of adding an boolean settng to the sd_bus object, and another one to sd_device and so on – maybe instead add one to log.c called "log_request_context_requested" or so that sd-bus, sd-device and so on all check on their own. i.e. a single global option I think is much nicer than having to enable this in every subsystem individually.

poettering replied:
or maybe make log_request_context_requested a tristate:

1. if 0: no extended log contexts are added anywhere
2. if > 0: extended log contexts are added everywhere
3. if < 0: extended log contexts are done if debug logging is on

< 0 would be the default of course.

And then maybe add an env var for this too, so that callers can request this from the command lines explicitly without necessarily having to enable debug logging.

poettering commented:
looks good, but I think the logic should be enabled differently, see my comments above.
